### DISCUSSION TITLE
Enabling beans conditionally during runtime

### DISCUSSION THREAD

**Discussion by ctron** [quarkusio/quarkus#29831] [Q&A]
I want to enable beans during runtime, based one some property:

```java
@Startup
@OnlyIfPropertyIs(property="foo.bar.type", value="baz")
class MyStuff {
}
```

And:

```java
interface Foo {}

@OnlyIfPropertyIs(property="foo.bar.type", value="baz")
@ApplicationScoped
class Bar implements Foo {}

@Default
@ApplicationScoped
class Baz implements Foo {}

class MyOtherStuff {
  @Inject
  Foo foo;
}
```

I found `@LookupUnlessProperty`. However, to my understanding, these only work with "lookups", so I have to use `Instance<Foo>`. I would be ok with this, but this doesn't seem to work with the `@Startup` type bean.

**Comment by r00ta** [comment]
Hi @ctron , In our project we faced the same challenge multiple times. You can find here one of the many examples we used as workaround https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/networking/NetworkingServiceProducer.java . In short, we add a `Producer` bean that injects the config property and then `@Produces` a bean accordingly.

**Reply by ctron** [reply]
This is indeed a nice pattern. But it looks like as if you no longer can use CDI to configure those returned instances. But you need to instantiate them manually, with all their dependencies.

**Comment by mkouba** [comment]
> I want to enable beans during runtime, based one some property:

There's no way to dynamically include/exclude beans at runtime. Quarkus does resolve all injection points (dependencies) at build time, i.e. the set of available beans cannot be modified when the app is built. This allows for various optimizations that follow the core design ideas of Quarkus.

However, you can enable/disable beans during the build, i.e. for a [specific build profile](https://quarkus.io/guides/cdi-reference#enable_build_profile) and [when a build time property has/has not a specific value](https://quarkus.io/guides/cdi-reference#enable_build_properties).

> I found `@LookupUnlessProperty`. However, to my understanding, these only work with "lookups", so I have to use `Instance<Foo>`. I would be ok with this, but this doesn't seem to work with the `@Startup` type bean.

Exactly,  [`@LookupIfProperty` and `@LookupUnlessProperty`](https://quarkus.io/guides/cdi-reference#declaratively-choose-beans-that-can-be-obtained-by-programmatic-lookup) can only be used to affect the programmatic lookup.

**Reply by bsdelf** [reply]
Sad to hear that. I thought `LookupIfProperty ` could be used to enable/disable specific router at runtime, turns out it never worked.

```
@LookupIfProperty(name = "service.flags.book", stringValue = "true")
@Path("/api/books")
public class BookResource {
}
```

**Reply by mkouba** [reply]
Nope, `@LookupIfProperty` and `@LookupUnlessProperty` can only be used for [programmatic lookup](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#programmatic_lookup).

**Reply by sixcorners** [reply]
Would this make more sense if this said "build time" instead of "runtime"?  
https://quarkus.io/guides/cdi-reference#declaratively-choose-beans-that-can-be-obtained-by-programmatic-lookup

**Reply by mkouba** [reply]
No, the programmatic lookup is a runtime feature and both `@LookupIfProperty` and `@LookupUnlessProperty` do check a runtime configuration property.
